### PR TITLE
Optimize logistic regression plates 3, 3 prime, and 5

### PR DESCRIPTION
**Pull request by MattsonCam**
In this pr, a random search of the regularization parameter is performed using data from plates 3, 3 prime, and 5. The final model is then trained on all of the cross-validated data and saved. The folds and data splits used for optimization were also saved for later evaluation of training and validation data. Please let me know if you have any suggestions.

### CODE REVIEW

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:65]: Unless there's especially a good reason for it (such as namespace overloading, etc) consider moving this block up near the other imports to help keep things organized in this notebook.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:104]: Consider using [jupyterlab_code_formatter](https://github.com/ryantam626/jupyterlab_code_formatter) or [nbqa-black](https://nbqa.readthedocs.io/en/latest/tutorial.html#run-nbqa-black) (note: there's a [pre-commit check](https://nbqa.readthedocs.io/en/latest/pre-commit.html) too!) for formatting your Python code within a notebook.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:90]: Consider adding a short sentence to describe what this and the other functions in this file do.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]:
```suggestion
    # Combine parameters in current search with logistic regression parameters
```

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]:
```suggestion
datasplitsdf.to_parquet(f"{data_path}/scv_folds_{data_suf}.parquet")
```

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:164]: I love how you've broken up the sections of this notebook with headers and descriptions like this one. Great job!

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]:
```suggestion
# We achieve this by selecting specific holdout wells that maximize the minority class in the train-validation set.
```

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]: Minor, just trying to avoid repeated use of "that".
```suggestion
# In other words, we choose the combination of wells for train-validation that, together, include the highest number of cells in the genotype category which has the fewest number of cells.
```

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]: Double checking, should this be stored in the model or perhaps a "methods" dir to help distinguish between data? Feel free to dismiss if this doesn't add value.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:46]: Seeing how this block repeats in multiple notebooks, would it make sense to create a reusable module + related function to perform this work? If you're able to, I suggest practicing DRY to help increase your development velocity with this. From an implementation standpoint, it could be that a gradual approach would work well (adding the use of a function as new notebooks are created).

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:93]: Consider adding type hints suggested here to the function definition (e.g. `def down_sample_by_genotype(_df: pd.DataFrame):`). This comment applies to all the functions + their parameters found in this file.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:333]: Consider the use of [`sklearn`'s pipelines](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) to help streamline the creation and implementation aspects of models.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:258]: Consider using `dict.items()` to help specify named elements of the dictionary within a comprehension.

For example:

```python
example = {"one": "cat", "two": "dog"}
example = {key: val for key, val in example.items()}
```

**Review by d33bs — CHANGES_REQUESTED** [review]
Nice job! I'm requesting changes mostly due to what looked like a misalignment in file naming for a parquet export labeled as a `.joblib`. Once that's in place, please feel free to re-request a review. All other comments are up for your consideration. Please don't hesitate to let me know if you have any questions along the way.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:65]: I would, but I have to append the path using the root_dir variable before I can import this

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:104]: I think I'm already using black, but I'll try rerunning precommit.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:90]: Thanks for the suggestion!

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]: Changed

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]: Good catch, changed

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:164]: Thanks!

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]: Changed

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]: Changed

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]: It could, but I think the model would also fall under this category of "objects saved containing subroutines". I don't think there is a rigid distinction here

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:46]: I've thought about this, but then I would have to specify a relative path to this new module, in which case, It would be easier to just to specify the relative path to the root git directory and avoid this using this code.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:93]: Is there any tool that will add these type hints in code for python?

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:333]: I've thought about this. I like the idea, and it may be helpful for understanding the code as well. I think you can also create your own functions for these pipelines. For now, it has been quicker to create the custom functions I need for specific analysis.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:258]: Thanks for the suggestion, this can be very helpful. Here, I just wanted to make code that could account for multiple hyperparameters for use in future analysis, but it may be easier to read that way.

**Comment by MattsonCam** [comment]
Hey @d33bs, actually, I may add something to this pr, so could you wait to provide a re-review?

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:104]: I reran my precommit

**Comment by d33bs** [comment]
Thanks @MattsonCam! I'll wait for your go-ahead here on a re-review.

**Comment by MattsonCam** [comment]
> Thanks @MattsonCam! I'll wait for your go-ahead here on a re-review.

Thanks for waiting @d33bs, it's ready now!

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]: Consider using the `numpy` arrays without converting to Python lists through [`np.concatenate()`](https://numpy.org/doc/stable/reference/generated/numpy.concatenate.html). It also might be beneficial to consider using [`pd.concat()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) (creating dataframes on each iteration, e.g. `pd.DataFrame({"value_1":np.array, "value_2":np.array})`). The benefit here might be that you are able to make the syntax more declarative and avoid translation from `numpy` to Python objects (dropping the calls to `tolist()`). For larger objects, there could be performance implications here too, but I'm unsure off-hand which would offer a better outcome. Mostly my suggestion would be to avoid data translation if you can avoid it.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:341]: Double checking: would `eval_data` at this point still include the prior data which was extended to it above? If so, does this impact any analysis which would be performed on it?

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (d33bs) [inline 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:302]: Consider the use of [`np.repeat()`](https://numpy.org/doc/stable/reference/generated/numpy.repeat.html) here.

**Review by d33bs — APPROVED** [review]
Nice job! I left a few comments which might need a look but overall thought things looked good. Please don't hesitate to let me know if you have any questions.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py]: I've thought about this, I may have to do some benchmarking at some point to determine which is faster. I think appending to lists is faster compared to appending to a numpy array, but I'm not sure about the data translation piece.

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:341]: It would, yes, this was intentional

File: 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py
Comment (MattsonCam) [reply 1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py:302]: This could be beneficial for decreasing runtime, I may have to look into this along with the above data translation suggestion.

**Comment by MattsonCam** [comment]
Thanks @d33bs, merging now!

**Pull request merged**
